compiler/frontend/src/org/jetbrains/kotlin/idea/MainFunctionDetector.kt

```diff
@@ -178,8 +176,11 @@ class MainFunctionDetector {
         return null
     }
 
-    private fun findMainFunction(declarations: List<KtDeclaration>) =
-        declarations.filterIsInstance<KtNamedFunction>().find { isMain(it) }
+    fun findMainFunction(declarations: List<KtElement>): KtNamedFunction? =
+        declarations
+            .ifEmpty { return@findMainFunction null }
+            .filterIsInstance<KtNamedFunction>().find { isMain(it) }
+            ?: findMainFunction(declarations.flatMap { it.getChildrenOfType<KtElement>().toList() })
```

Comment:
I suggest to rewrite it a bit, so that in case there are several `main`s in the same file, we would avoid writing any of them to the manifest (might be worth adding a test for this case). Just like we do when several files declare top-level `main`.
```
    fun findMainFunction(container: KtDeclarationContainer): KtNamedFunction? =
        container.declarations.mapNotNull { declaration ->
            when (declaration) {
                is KtNamedFunction -> declaration.takeIf(mainFunctionDetector::isMain)
                is KtDeclarationContainer -> findMainFunction(declaration)
                else -> null
            }
        }.singleOrNull()
```

Reply:
Great point, thanks!